src/vision/status_quo/barbara_builds_an_async_executor.md

```diff
@@ -0,0 +1,161 @@
+
+First, Barbara found [crossbeam-deque](https://crates.io/crates/crossbeam-deque) provides work-stealing deques of good quality. She could use it to build task schedulers. It is not the main part of the story. Each working thread can get a task from the scheduler and poll it in loop.
+
+But wait, what is a "task"?
+
+At first, Barbara thought it must contain the `Future` itself and the additional priority which was used by the scheduler. So she first wrote:
+
+```rust
+pub struct Task {
+    future: Pin<Box<dyn Future<Output = ()> + Send + 'static>>,
+    priority: u8
+}
+```
+
+And the working thread loop should run something like:
+
+```rust
+pub fn poll_task(task: Task) {
+    let waker = todo!();
+    let mut cx = Context::from_waker(&waker);
+    task.future.as_mut().poll(&mut cx);
+}
+```
+
+"How to create a waker?" Barbara asked herself. Quickly, she found the `Wake` trait. After reading the documentation, she realized the task in the scheduler should be stored in an `Arc`. And to implement `Wake`, the `Task` should also contain the sender of the scheduler. She changed the code to something like this:
+
+```rust
+pub struct Task {
+    future: Pin<Box<dyn Future<Output = ()> + Send + 'static>>,
+    scheduler: SchedulerSender,
+    priority: u8,
+}
+
+unsafe impl Sync for Task {}
+
+impl Wake for Task {
+    fn wake(self: Arc<Self>) {
+        self.scheduler.send(self.clone());
+    }
+}
+
+pub fn poll_task(task: Arc<Task>) {
+    let waker = Waker::from(task.clone());
+    let mut cx = Context::from_waker(&waker);
+    task.future.as_mut().poll(&mut cx);
+//  ^^^^^^^^^^^ cannot borrow as mutable
+}
+```
+
+The code still needed some change because the `future` in the  `Arc<Task>` became immutable. 
+
+"Okay. I can guarantee `Task` is created from a `Pin<Box<Future>>` and I think the same future will not be polled concurrently in two threads. So let me bypass the safety checks." Barbara confidently used some `unsafe` blocks to make it compile.
+
+```rust
+pub struct Task {
+    future: *mut (dyn Future<Output = ()> + Send + 'static),
+    ...
+}
+
+unsafe impl Send for Task {}
+unsafe impl Sync for Task {}
+
+pub fn poll_task(task: Arc<Task>) {
+    ...
+    unsafe {
+        Pin::new_unchecked(&mut *task.future).poll(&mut cx);
+    }
+}
+```
+
+Luckily, a colleague of Barbara noticed something wrong. The `wake` method could be called multiple times so multiple copies of the task could exist in the scheduler. The scheduler might not work correctly because of it and a more severe problem was that multiple threads might get copies of the same task from the scheduler and cause a race in polling the future.
+
+Barbara soon got a idea to solve it. She added a state field to the `Task`. By carefully maintaining the state of the task, she could guarantee there are no duplicate tasks in the scheduler and no race can happen when polling the future:
```

Comment:
Just as a suggestion, maybe this paragraph can say something about how she spends a lot of effort in writing the code below (at least, it would take me a lot of effort). Then the reader is already focused on the pain of having to write this to satisfy such a simple requirement.